fdbserver/GrvProxyServer.actor.cpp

```diff
@@ -808,6 +837,7 @@ ACTOR static Future<Void> transactionStarter(GrvProxyInterface proxy,
 			// transactionQueue->span.swap(span);
 
 			auto& req = transactionQueue->front();
+			req.queueIterations++;
```

Comment:
If I understand correctly, this means that we will consider ratekeeper to be throttling if the same entry is sitting at the front of the queue for multiple passes of this logic. Then that same entry must sit there for some duration of time in order for us to trigger the throttled state, otherwise the next entry in the queue could cause us to clear the state.

I wonder if this might be too permissive in that it requires RK to completely lock down for some duration. It could be that ratekeeper is releasing transactions at a low rate compared to the number coming in, so only occasional transactions see this incremented more than once. The current knob value of 2 iterations is also a special case here in that it will still activate sometimes, but the current behavior of this is very dependent on that value.

Reply:
To clarify, my concern is not so much that we the fact that a request gets held in the queue as a signal that it is throttled. Rather, I'm nervous that since we only update this state for the first element in the queue, if we get any allocation that allows more than one transaction it will immediately reset the state to unthrottled. This is because we not only require a request to report that it was throttled, but that state must hold for an additional 0.1 seconds before we notify the client.

If any request gets started in that intervening 0.1 seconds without sitting at the front of the queue for multiple iterations, then we reset the clock. This is true if that subsequent transaction is an immediate priority transaction, or if it is a normal priority transaction while batch transactions are getting throttled, or even if it spent multiple iterations in the queue deeper than the first element.

Reply:
If we want to use the fact that the queue wasn't exhausted as a signal for throttling, probably the most direct way to do that would be to measure the size of the queues at the end of each pass through the transaction starter.

We should probably just ignore the immediate queue, and then for the normal and batch queues we would mark the proxy as throttling if the queue was non-empty. If that state persisted for your specified duration, then you could notify the client.

To avoid interactions between the priorities, we should have a signal for each. You could probably send the signal associated with the priority of each request, but you would need to track both priorities separately on the client so that you could make separate decisions for each.

Reply:
I see, because we only look at the front of the queue, any of the examples you gave will reset our grace period if it follows immediately after our original front is processed. Regarding immediate priority transactions, I added a change so that they will not be counted when checking for consecutive throttled messages, which I hope is fine. For the other 2 examples, is the idea that we should be more granular and track normal and batch separately? Or should we instead be more aggressive and apply the iteration count to all items in queue instead of just the front?

Reply:
First, let's assume there are no priorities as that will make my description simpler.

Rather than requiring that every entry be at the front of the queue for 2+ requests over 0.1 seconds, we could change it so that we consider the proxy throttling if the queue is not emptied. Each time you stop the transaction starter loop with a non-empty queue, some entry is going to have its queue iterations incremented, but we don't have to worry about the case where we are able to start 2+ requests in one go and some intermediate request had only one iteration.

In essence, what you are tracking is how long it has been since the last time you managed to flush the request queue. If it's been longer than 0.1s since you've done so, then you can say you are throttled. Therefore you don't even have to track queue iterations explicitly, you just need to check what the queue size was when you finished the loop.

Now if we factor priorities back into this, each of the queues is going to have different states. We should not track the immediate queue at all, and instead always send back that we aren't throttled. For normal and batch requests, we would track how long the associated queue has been non-empty and send back a throttled response when that has been long enough.